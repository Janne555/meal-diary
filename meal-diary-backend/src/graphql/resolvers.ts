import { Resolvers, RoleResolvers, UserResolvers } from "../types/generated"
import { Context } from "../types"
import { UnauthorizedRequest } from '../errors'
import core from '@actions/core'


const resolvers: Pick<Resolvers<Context>, 'Mutation' | 'Query'> = {
  Query: {
    users: (_, __, { userDataSource }) => {
      if (userDataSource) {
        return userDataSource.getUsers()
      } else {
        throw new UnauthorizedRequest
      }
    },
    foods: async (_, { params }, { foodDataSource }) => {
      throw Error("wtf")
      console.error("at foods resolver")
      if (foodDataSource) {
        console.error("datasource available")
        return foodDataSource.getFoods(params ?? {})
      } else {
        console.error("not authorized")
        throw new UnauthorizedRequest
      }
    },
    config: (_, __, { configDataSource }) => {
      if (configDataSource) {
        return configDataSource.getConfig()
      } else {
        throw new UnauthorizedRequest
      }
    },
    food: (_, { id }, { foodDataSource }) => {
      if (foodDataSource) {
        return foodDataSource.getFood(id)
      } else {
        throw new UnauthorizedRequest
      }
    }
  },
  Mutation: {
    assignRoleToUser: async (_, { roleIds, userId }, { userDataSource }) => {
      if (userDataSource) {
        await userDataSource.assignRolestoUser({ id: userId }, { roles: roleIds })
        return userDataSource.getUser({ id: userId })
      } else {
        throw new UnauthorizedRequest
      }
    },
    updateConfig: async (_, params, { configDataSource }) => {
      if (configDataSource) {
        try {
          const config = await configDataSource.updateConfig(params.config)
          return {
            code: "200",
            success: true,
            message: "Config updated succesfully",
            config
          }
        } catch (error) {
          const config = await configDataSource.getConfig()
          return {
            code: "500",
            success: false,
            message: `Failed to update config: ${error.message}`,
            config
          }
        }
      } else {
        throw new UnauthorizedRequest
      }

    },
    addFood: async (_, { food }, { foodDataSource }) => {
      if (foodDataSource) {
        const storedFood = await foodDataSource.addFood(food)
        return {
          code: "",
          message: "",
          success: true,
          food: storedFood
        }
      } else {
        throw new UnauthorizedRequest
      }
    }
  }
}

const Role: Partial<RoleResolvers<Context>> = {
  permissions: async (r, _, { userDataSource }) => {
    if (userDataSource) {
      return r.id ? (await userDataSource.getPermissionsInRole({ id: r.id })) : null
    } else {
      throw new UnauthorizedRequest
    }
  }
}

const User: Partial<UserResolvers<Context>> = {
  roles: async (u, _, { userDataSource }) => {
    if (userDataSource) {
      return u.userId ? (await userDataSource.getUserRoles({ id: u.userId })) : null
    } else {
      throw new UnauthorizedRequest
    }
  },
  permissions: async (u, _, { userDataSource }) => {
    if (userDataSource) {
      return u.userId ? (await userDataSource.getUserPermissions({ id: u.userId })) : null
    } else {
      throw new UnauthorizedRequest
    }
  }
}

export default {
  ...resolvers,
  Role,
  User
}