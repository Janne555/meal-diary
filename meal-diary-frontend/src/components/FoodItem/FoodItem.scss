.food-item {
  > div {
    margin: 1em;
    display: flex;
    flex-wrap: wrap;
  }
}
