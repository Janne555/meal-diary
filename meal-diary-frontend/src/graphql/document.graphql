query getFoodNames {
  foods {
    id
    name {
      fi
      en
    }
  }
}

query getFood($id: ID!) {
  food(id: $id) {
    id
    name {
      fi
      en
    }
    enerc
    fat
    prot
    sugar
    fibc
    nacl
  }
}