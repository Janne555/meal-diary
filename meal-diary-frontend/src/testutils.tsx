import React from 'react'
import { render } from "@testing-library/react"
import { MockedProvider, MockedProviderProps } from '@apollo/react-testing'

// todo add auth provider mock
const createContextfulRenderer = (mockedProviderProps: MockedProviderProps = {}) => (element: JSX.Element) => render(
  <React.StrictMode>
    <MockedProvider addTypename={false} {...mockedProviderProps} >
      {element}
    </MockedProvider>
  </React.StrictMode>,
)

export {
  createContextfulRenderer
}