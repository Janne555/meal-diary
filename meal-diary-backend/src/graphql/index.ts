import { ApolloServer } from "apollo-server-express";
import { DIRECTIVES } from "@graphql-codegen/typescript-mongodb";
import { UserToken } from "../types";
import { AuthDirective } from "./directives";
import resolvers from "./resolvers";
import typeDefs from './schema';
import { isToken } from "../utils/typeguards";
import { UserDataSource, FoodDataSource } from "../datasources";
import ConfigDataSource from "../datasources/configDataSource";

const graphqlServer = new ApolloServer({
  typeDefs: [DIRECTIVES].concat(typeDefs), resolvers: { ...resolvers }, context: ({ req }) => {
    const userToken = ((): UserToken | undefined => {
      if (isToken(req.user)) {
        return {
          ...req.user,
          hasRoles(roles: string[]): boolean {
            return roles.every(role => this.permissions.includes(role))
          }
        }
      } else {
        return undefined
      }
    })()
    if (userToken) {
      return {
        userDataSource: new UserDataSource(),
        foodDataSource: new FoodDataSource(userToken),
        configDataSource: new ConfigDataSource(),
        userToken
      }
    } else {
      return {}
    }
  },
  schemaDirectives: {
    auth: AuthDirective
  }
});

export default graphqlServer