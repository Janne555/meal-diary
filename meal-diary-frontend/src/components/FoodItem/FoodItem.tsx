import React from 'react'
import { useGetFoodQuery } from '../../graphql/generated'
import { CircularProgress } from '@material-ui/core'
import NutritionalComponent from '../NutritionalComponent/NutritionalComponent'
import Typography from '@material-ui/core/Typography';

type Props = {
  id: string
}

function FoodItem({ id }: Props) {
  const { loading, data } = useGetFoodQuery({ variables: { id } })

  if (loading) {
    return <CircularProgress />
  }

  return (
    <div className="food-item">
      <Typography variant="h1">{data?.food?.name.fi}</Typography>
      <div>
        {
          Object.entries(data?.food ?? {})
            .filter(([key]) => !["__typename", "name"].includes(key))
            .map(([key, value]) => (
              <NutritionalComponent key={key} name={key} value={`${value}`} />
            ))
        }
      </div>
    </div>
  )
}

export default FoodItem