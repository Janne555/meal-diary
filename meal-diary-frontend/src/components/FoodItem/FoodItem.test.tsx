import React from 'react'
import FoodItem from './FoodItem'
import { createContextfulRenderer } from '../../testutils';
import { waitForDomChange } from '@testing-library/react';
import { GetFoodDocument, GetFoodQuery } from '../../graphql/generated'



describe('<FoodItem />', () => {
  it('should match snapshot', async () => {
    const result: { data: GetFoodQuery } = {
      data: {
        food: {
          id: "foo",
          name: {
            en: "foodname",
            fi: "ruokanimi"
          },
          enerc: null,
          fat: null,
          prot: null,
          fibc: null,
          nacl: null,
          sugar: null
        }
      }
    }

    const mocks = [
      {
        request: {
          query: GetFoodDocument,
          variables: {
            id: 'foo',
          },
        },
        result
      },
    ]

    const render = createContextfulRenderer({ mocks })
    const { asFragment } = render(<FoodItem id="foo" />)
    await waitForDomChange()
    expect(asFragment()).toMatchSnapshot()
  });
});