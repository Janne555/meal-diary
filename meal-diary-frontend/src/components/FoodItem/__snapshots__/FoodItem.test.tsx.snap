// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FoodItem /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="food-item"
  >
    <h1
      class="MuiTypography-root MuiTypography-h1"
    >
      ruokanimi
    </h1>
    <div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            id
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            foo
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            enerc
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            fat
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            prot
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            sugar
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            fibc
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root nutritional-component MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div>
          <img
            alt="foo"
            src="images/fuel.svg"
          />
          <h1
            class="MuiTypography-root MuiTypography-body1"
          >
            nacl
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            null
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
